pkg/helm/watcher/server.go

```diff
@@ -0,0 +1,94 @@
+package watcher
+
+import (
+	"io/ioutil"
+	"os"
+
+	sourcev1 "github.com/fluxcd/source-controller/api/v1beta1"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	_ "k8s.io/client-go/plugin/pkg/client/auth"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/healthz"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+
+	"github.com/weaveworks/weave-gitops/pkg/helm"
+	//+kubebuilder:scaffold:imports
+	"github.com/weaveworks/weave-gitops/pkg/helm/watcher/cache"
+	"github.com/weaveworks/weave-gitops/pkg/helm/watcher/controller"
+)
+
+var (
+	scheme   = runtime.NewScheme()
+	setupLog = ctrl.Log.WithName("setup")
+)
+
+type Watcher struct {
+	cache       cache.Cache
+	repoManager *helm.RepoManager
+}
+
+func NewWatcher(kubeClient client.Client, cache cache.Cache) (*Watcher, error) {
+	tempDir, err := ioutil.TempDir("", "helmrepocache")
+	if err != nil {
+		return nil, err
+	}
+
+	return &Watcher{
+		cache:       cache,
+		repoManager: helm.NewRepoManager(kubeClient, tempDir),
+	}, nil
+}
+
+func (w *Watcher) StartWatcher() error {
+	utilruntime.Must(clientgoscheme.AddToScheme(scheme))
+	utilruntime.Must(sourcev1.AddToScheme(scheme))
+	ctrl.SetLogger(zap.New(zap.UseFlagOptions(&zap.Options{
+		Development: true,
+	})))
+
+	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
+		Scheme:                 scheme,
+		MetricsBindAddress:     ":9980",
```

Comment:
These should probably be configured from a CLI. Or maybe we don't care for now. I'm not sure. :) Thoughts?

Reply:
You may want to consider using `fluxcd/pkg/runtime` as it contains helpers for metrics, events, logging, watching sources, sending alerts to notification-service and many more.

Reply:
Hey @stefanprodan! I actually looked at this package! :) And, please forgive me, I'm probably blind, but which part of this package is supposed to helm with watching sources? Do you mean the predicate for example? I'm not sure how I would have used that. I know that the metrics part is probably interesting for me, but to be honest, this is such a small thing, I'm thinking about completely leaving out that part. WDYT?

Reply:
Here is an example on how to setup structured logging, API rate limits, metrics, events, leader election, health checks: https://github.com/fluxcd/kustomize-controller/blob/main/main.go#L83-L107

Reply:
Thanks!